core/slice.go

```diff
@@ -647,29 +647,55 @@ func (sl *Slice) init(genesis *Genesis) error {
 		rawdb.WriteTermini(sl.sliceDb, genesisHash, genesisTermini)
 
 		// Append each of the knot blocks
-
-		knot := genesis.Knot[:]
-		for _, block := range knot {
-			sl.AddPendingEtxs(types.PendingEtxs{block.Header(), emptyPendingEtxs})
-			if block != nil {
-				location := block.Header().Location()
-				if nodeCtx == common.PRIME_CTX {
-					sl.bestPhKey = block.Hash()
-					rawdb.WriteBody(sl.sliceDb, block.Hash(), block.NumberU64(), block.Body())
-					_, err := sl.Append(block.Header(), types.EmptyHeader(), genesisHash, block.Difficulty(), false, false, nil)
-					if err != nil {
-						log.Warn("Failed to append block", "hash:", block.Hash(), "Number:", block.Number(), "Location:", block.Header().Location(), "error:", err)
-					}
-				} else if location.Region() == common.NodeLocation.Region() && len(common.NodeLocation) == common.REGION_CTX {
-					sl.bestPhKey = block.Hash()
-					rawdb.WriteBody(sl.sliceDb, block.Hash(), block.NumberU64(), block.Body())
-				} else if bytes.Equal(location, common.NodeLocation) {
-					sl.bestPhKey = block.Hash()
-					rawdb.WriteBody(sl.sliceDb, block.Hash(), block.NumberU64(), block.Body())
+		sl.bestPhKey = genesisHash
+
+		sl.hc.SetCurrentHeader(genesisHeader)
+		manifest := types.BlockManifest{genesisHash}
+		manifestHash := types.DeriveSha(manifest, trie.NewStackTrie(nil))
+		sl.AddPendingEtxs(types.PendingEtxs{Header: genesisHeader, Etxs: []types.Transactions{}})
+
+		pendingHeader := types.EmptyHeader()
+		for index := 0; index < common.HierarchyDepth; index++ {
+			pendingHeader.SetParentHash(genesisHash, index)
+			pendingHeader.SetNumber(big.NewInt(1), index)
+			pendingHeader.SetBaseFee(misc.CalcBaseFee(sl.config, genesisHeader), index)
+			pendingHeader.SetGasLimit(params.MinGasLimit, index)
+			pendingHeader.SetGasUsed(uint64(0), index)
+			pendingHeader.SetDifficulty(sl.hc.genesisHeader.Difficulty(index), index)
+			if index == common.PRIME_CTX {
+				pendingHeader.SetRoot(common.HexToHash("0x75c5a24ee5d9cca1a7d47b9f65d2ac9cf92c5ad42f1fddea02299aa05ce4efa2"), index)
+				pendingHeader.SetCoinbase(common.HexToAddress("0x00114a47a5d39ea2022dd4d864cb62cfd16879fc"), index)
+			} else if index == common.REGION_CTX {
+				switch common.NodeLocation.Region() {
+				case 0:
+					pendingHeader.SetRoot(common.HexToHash("0x6b3bb2136ece6e2dd7e65990b05d51b6e0911882e448d277e2944ce378497c62"), index)
```

Comment:
I would prefer not to see magic numbers, especially for root hashes. The first pending header should be treated the same as the genesis block: the genesis parameters are hard-coded, but the block itself is generated in plain view from those parameters.

I don't mind hardcoding some parameters if it helps (but they should be in params/config.go or something similar to how genesis is handled), but the pending header itself (especially the root hashes) should be computed in a clearly verifiable way.

Alternatively, if you want to implement the method to asynchronously get a pending header from a block, you could use that method here.

Reply:
Note, if you go with the latter approach, there is no need for you `FinalizeAtContext()` method, which makes me itchy for other reasons.

Reply:
All these logics are gone, and state functions are back to where they were